src/vs/workbench/contrib/notebook/browser/contrib/notebookVariables/notebookInlineVariables.ts

```diff
@@ -0,0 +1,356 @@
+	}
+}
+
+export class NotebookInlineVariablesController extends Disposable implements INotebookEditorContribution {
+
+	static readonly id: string = 'notebook.inlineVariablesController';
+
+	private cellDecorationIds = new Map<ICellViewModel, string[]>();
+	private cellContentListeners = new ResourceMap<IDisposable>();
+
+	constructor(
+		private readonly notebookEditor: INotebookEditor,
+		@INotebookKernelService private readonly notebookKernelService: INotebookKernelService,
+		@INotebookExecutionStateService private readonly notebookExecutionStateService: INotebookExecutionStateService,
+		@ILanguageFeaturesService private readonly languageFeaturesService: ILanguageFeaturesService,
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IDebugService private readonly debugService: IDebugService,
+	) {
+		super();
+
+		this._register(this.notebookExecutionStateService.onDidChangeExecution(async e => {
+			if (!this.configurationService.getValue<boolean>(NotebookSetting.notebookInlineVariables)) {
+				return;
+			}
+
+			if (this.isCellExecutionStateChangedEvent(e)) {
+				await this.updateInlineVariables(e);
+			}
+		}));
+	}
+
+	private async updateInlineVariables(event: ICellExecutionStateChangedEvent): Promise<void> {
+		if (this.debugService.state !== 0) { // debug session is a state other than inactive, so clear and defer to whatever it's handling
+			this._clearNotebookInlineDecorations();
+			return;
+		}
+
+		if (this.notebookEditor.textModel?.uri !== event.notebook) {
+			return;
+		}
+
+		if (event.changed) { // undefined -> execution was completed, so return on all else
+			return;
+		}
+
+		const cell = this.notebookEditor.getCellByHandle(event.cellHandle);
+		if (!cell || cell.language !== 'python') { // TODO: support more than just python
+			return;
+		}
+		const model = cell.textModel;
+		if (!model) {
+			return;
+		}
+
+		this.clearCellInlineDecorations(cell);
+		const inlineDecorations: IModelDeltaDecoration[] = [];
+
+
+		// might need to be a new provider registry, since we will have these that don't necessarily provide at the right timing
+		if (this.languageFeaturesService.inlineValuesProvider.has(model)) {
+			// use extension based provider, borrowed from https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugEditorContribution.ts#L679
+			const lastLine = model.getLineCount();
+			const lastColumn = model.getLineMaxColumn(lastLine);
+			const ctx: InlineValueContext = {
+				frameId: 0, // ignored, we won't have a stack from since not in a debug session
+				stoppedLocation: new Range(lastLine, lastColumn, lastLine, lastColumn) // executing cell by cell, so "stopped" location would just be the end of document
+			};
+
+			const providers = this.languageFeaturesService.inlineValuesProvider.ordered(model).reverse();
+			const lineDecorations = new Map<number, InlineSegment[]>();
+
+			const fullCellRange = new Range(1, 1, lastLine, lastColumn);
+
```

Comment:
maybe we can early return if there is no active kernel or the kernel doesn't have variable provider.

Reply:
I think just continuing is likely safer, as there's a possibility of the providers sending variableText alongside lookups, in which case returning early would get rid of the chance to display those. We do return early if there's no variables provided, so that case would be adressed.